src/blocks/weather.rs

```diff
@@ -185,24 +187,86 @@ pub async fn run(config: toml::Value, mut api: CommonApi) -> Result<()> {
+            loop {
+                let location: Option<Coordinates> = api.recoverable(find_ip_location).await?;
+                let data = api
+                    .recoverable(|| async { provider.get_weather(location).await })
+                    .await?;
+                widget.set_icon(data.icon.to_icon_str())?;
+                widget.set_values(data.into_values());
+                api.set_widget(&widget).await?;
 
-        select! {
-            _ = sleep(config.interval.0) => (),
-            _ = api.wait_for_update_request() => (),
+                select! {
+                    _ = sleep(config.interval.0) => (),
+                    _ = api.wait_for_update_request() => (),
+                }
+            }
+        } else {
+            // Two timers, one to rerender the block and the other to update the location.
+            let mut interval = config.interval.timer();
+            let mut autolocate_interval = autolocate_interval.timer();
+
+            // Initial pass
+            let mut location: Option<Coordinates> = api.recoverable(find_ip_location).await?;
+            let data = api.recoverable(|| provider.get_weather(location)).await?;
+            widget.set_icon(data.icon.to_icon_str())?;
+            widget.set_values(data.into_values());
+            api.set_widget(&widget).await?;
+
+            loop {
+                select! {
+                    biased; // if both timers `tick()` autolocate should run first
+                    _ = autolocate_interval.tick() => {
+                        location = api.recoverable(find_ip_location).await?;
+                    }
+                    _ = interval.tick() => {
+                        let data = api
+                            .recoverable(|| provider.get_weather(location))
+                            .await?;
+                        widget.set_icon(data.icon.to_icon_str())?;
+                        widget.set_values(data.into_values());
+                        api.set_widget(&widget).await?;
+                    },
+                    // On update request autolocate and update the block.
+                    _ = api.wait_for_update_request() => {
+                        location = api.recoverable(find_ip_location).await?;
+
+                        let data = api
+                            .recoverable(|| provider.get_weather(location))
+                            .await?;
+                        widget.set_icon(data.icon.to_icon_str())?;
+                        widget.set_values(data.into_values());
+                        api.set_widget(&widget).await?;
+
+                        // both intervals should be reset after a manual sync
+                        autolocate_interval.reset();
+                        interval.reset();
+                    },
+                }
+            }
+        }
+    } else {
+        loop {
+            let data = api
+                .recoverable(|| async { provider.get_weather(None).await })
```

Comment:
This one too.

Reply:
These would make good clippy suggestions